src/Symfony/Component/DependencyInjection/Compiler/AbstractRecursivePass.php

```diff
@@ -131,25 +132,35 @@ protected function getConstructor(Definition $definition, $required)
 
         if ($factory) {
             [$class, $method] = $factory;
+
+            if ('__construct' === $method) {
+                throw new RuntimeException(sprintf('Invalid service "%s": "__construct()" cannot be used as a factory method.', $this->currentId));
+            }
+
             if ($class instanceof Reference) {
-                $class = $this->container->findDefinition((string) $class)->getClass();
+                $factoryDefinition = $this->container->findDefinition((string) $class);
+                while ((null === $class = $factoryDefinition->getClass()) && $factoryDefinition instanceof ChildDefinition) {
+                    $factoryDefinition = $this->container->findDefinition($factoryDefinition->getParent());
+                }
             } elseif ($class instanceof Definition) {
                 $class = $class->getClass();
             } elseif (null === $class) {
                 $class = $definition->getClass();
             }
 
-            if ('__construct' === $method) {
-                throw new RuntimeException(sprintf('Invalid service "%s": "__construct()" cannot be used as a factory method.', $this->currentId));
-            }
-
             return $this->getReflectionMethod(new Definition($class), $method);
         }
 
-        $class = $definition->getClass();
+        while ((null === $class = $definition->getClass()) && $definition instanceof ChildDefinition) {
+            $definition = $this->container->findDefinition($definition->getParent());
+        }
 
         try {
             if (!$r = $this->container->getReflectionClass($class)) {
+                if (null === $class) {
```

Comment:
This is a minor DX improvement. If the class is null this message looks better to me than `Invalid service "foo": class "" does not exist.` (see `getReflectionMethod()` below where we already check if a class is set or not).